util/sha256/sha256.go

```diff
@@ -0,0 +1,144 @@
+// Package sha256 is like crypto/sha256,
+// but efficiently hashes fixed-width integers.
+package sha256
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"hash"
+)
+
+var _ hash.Hash = (*Hash)(nil)
+
+// Hash is a hash.Hash for SHA-256,
+// but has efficient methods for hashing fixed-width integers.
+type Hash struct {
+	// The optimization is to maintain our own block and
+	// only call h.Write with entire blocks.
+	// This avoids double-copying of buffers within sha256.digest itself.
+	// However, it does mean that sha256.digest.x goes unused,
+	// which is a waste of 64B.
+
+	h  hash.Hash              // always *sha256.digest
+	x  [sha256.BlockSize]byte // equivalent to sha256.digest.x
+	nx int                    // equivalent to sha256.digest.nx
+}
+
+func New() *Hash {
+	return &Hash{h: sha256.New()}
+}
+
+func (h *Hash) Write(b []byte) (int, error) {
+	h.HashBytes(b)
+	return len(b), nil
+}
+
+func (h *Hash) Sum(b []byte) []byte {
+	if h.nx > 0 {
+		// This causes block mis-alignment. Future operations will be correct,
+		// but are less efficient until Reset is called.
+		h.h.Write(h.x[:h.nx])
+		h.nx = 0
+	}
+	return h.h.Sum(b)
+}
+
+func (h *Hash) Reset() {
+	h.h.Reset()
+	h.nx = 0
+}
+
+func (h *Hash) Size() int {
+	return h.h.Size()
+}
+
+func (h *Hash) BlockSize() int {
+	return h.h.BlockSize()
+}
+
+func (h *Hash) HashUint8(n uint8) {
+	// NOTE: This method is carefully written to be inlineable.
+	if h.nx <= len(h.x)-1 {
+		h.x[h.nx] = n
+		h.nx += 1
+	} else {
+		h.hashUint8Slow(n)
+	}
+}
+
+//go:noinline
```

Comment:
why?

Reply:
oh, because otherwise it blocks HashUint8 from being inlined?

Reply:
Yep...

Reply:
I'll add a comment to the effect.